File: server/src/platforms/Platform.ts
import type { ProgressUpdate } from "../progress.js";

export type MediaFormat = "audio" | "video";

export interface VideoInfo {
  title: string;
  duration: number;
  thumbnail: string | null;
  uploader: string | null;
}

export interface ConvertResult {
  id: string;
  filePath: string;
  title: string;
  ext: "mp3" | "mp4";
}

export interface PlaylistEntry {
  id: string;
  url: string;
  title: string;
  thumbnail: string | null;
  duration: number | null;
}

export interface PlaylistInfo {
  title: string;
  entries: PlaylistEntry[];
  /** Total number of videos in the playlist, so the UI knows whether more pages remain. */
  totalCount: number | null;
}

export interface UserFacingError {
  key: string;
  params?: Record<string, string | number>;
}

/**
 * One supported content source (YouTube, TikTok, ...). Every method here mirrors a function that
 * used to live in youtube.ts/validate.ts/utils.ts as a free function taking a URL — moved onto an
 * instance so each platform can override only what actually differs (see BasePlatform).
 */
export interface Platform {
  readonly id: string;
  /** Throws if this URL isn't actually usable by this platform. */
  checkAvailability(): void;
  fetchInfo(): Promise<VideoInfo>;
  buildFormatArgs(format: MediaFormat, quality: string): string[];
  download(format: MediaFormat, quality: string, onProgress: (update: ProgressUpdate) => void): Promise<ConvertResult>;
  /** Whether a failed download attempt is worth retrying, or is a known-permanent failure. */
  isRetryableError(err: unknown): boolean;
  describeError(err: unknown, fallbackRaw?: string): UserFacingError;
}

/** A platform that can also list a playlist's entries — not every platform has this concept. */
export interface PlaylistCapablePlatform extends Platform {
  fetchPlaylistInfo(start?: number, count?: number): Promise<PlaylistInfo>;
  /** Thumbnail to use for a playlist entry that yt-dlp didn't return one for, or null if there is none. */
  defaultThumbnail(entryId: string): string | null;
}

export function isPlaylistCapable(platform: Platform): platform is PlaylistCapablePlatform {
  return "fetchPlaylistInfo" in platform;
}

File: app/modules/ytdlp/android/src/main/java/expo/modules/ytdlp/platforms/Platform.kt
package expo.modules.ytdlp.platforms

import com.yausername.youtubedl_android.YoutubeDLRequest
import expo.modules.ytdlp.DownloadJob
import expo.modules.ytdlp.JobMetadata

/**
 * One supported content source (YouTube, TikTok, ...). Mirrors server/src/platforms/Platform.ts —
 * each method here used to be a function directly on DownloadQueue, keyed off the job's URL.
 */
interface Platform {
    /** Throws if this URL isn't actually usable by this platform. */
    fun checkAvailability()
    /** [cookiesPath] is the imported cookies.txt's absolute path (see CookiesStore), or null if none is stored. */
    fun fetchMetadata(job: DownloadJob, cookiesPath: String? = null): JobMetadata
    fun buildRequest(job: DownloadJob, outputTemplate: String, cookiesPath: String? = null): YoutubeDLRequest
    /**
     * The same options buildRequest() applies, as (flag, value) pairs — value is null for a bare
     * flag. The single source both buildRequest() and DownloadQueue's reproducible command log
     * draw from, so the two can never drift apart.
     */
    fun requestOptions(job: DownloadJob, outputTemplate: String, cookiesPath: String? = null): List<Pair<String, String?>>
    /** Whether a failed download attempt is worth retrying, or is a known-permanent failure. */
    fun isRetryableError(error: Throwable): Boolean
    fun describeError(error: Throwable): Pair<String, Map<String, Any?>?>
}

/** Applies a requestOptions()-style flag list to a YoutubeDLRequest for [url]. */
fun buildRequestFrom(url: String, options: List<Pair<String, String?>>): YoutubeDLRequest {
    val request = YoutubeDLRequest(url)
    options.forEach { (flag, value) -> if (value != null) request.addOption(flag, value) else request.addOption(flag) }
    return request
}

/** Appends --cookies <path> when a cookies file is stored; a no-op otherwise. Mirrors cookiesArgs() in server/src/cookies.ts. */
fun YoutubeDLRequest.withCookies(cookiesPath: String?): YoutubeDLRequest =
    if (cookiesPath != null) addOption("--cookies", cookiesPath) else this

/** How many playlist entries fetchPlaylistInfo() lists per call — mirrors PLAYLIST_PAGE_SIZE in server/src/environment.ts. */
const val PLAYLIST_PAGE_SIZE = 50

/** A platform that can also list a playlist's entries — not every platform has this concept. */
interface PlaylistCapablePlatform : Platform {
    suspend fun fetchPlaylistInfo(start: Int = 1, count: Int = PLAYLIST_PAGE_SIZE, cookiesPath: String? = null): Map<String, Any?>
    /** Thumbnail to use for a playlist entry that yt-dlp didn't return one for, or null if there is none. */
    fun defaultThumbnail(id: String): String?
}
